### PR TITLE
entity relationship model

### LINKED ISSUE #688: EntityRelationship Model

### PR DESCRIPTION
**Pull request by charleenfei**
<This is intended as a guide when describing-reviewing pull requests for go-centrifuge>
Close #688 .

### CODE REVIEW

File: documents/entityrelationship/model.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+	if data.TargetIdentity != "" {
+		if did, err := identity.NewDIDFromString(data.TargetIdentity); err == nil {
+			e.TargetIdentity = &did
+		} else {
+			return err
+		}
+	} else {
+		return identity.ErrMalformedAddress
+	}
+	e.Label = []byte(data.Label)
```
Comment (mikiquantum) [inline documents/entityrelationship/model.go]: We need to use here: 
```suggestion
	e.Label, err = hexutil.Decode(data.Label)
```

File: documents/entityrelationship/model.go [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+func (e *EntityRelationship) FromJSON(jsonData []byte) error {
+	return json.Unmarshal(jsonData, e)
+}
+
+// Type gives the EntityRelationship type.
+func (e *EntityRelationship) Type() reflect.Type {
+	return reflect.TypeOf(e)
+}
+
+// CalculateDataRoot calculates the data root and sets the root to core document.
```
Comment (mikiquantum) [inline documents/entityrelationship/model.go]:
```suggestion
// CalculateDataRoot calculates the data root.
```

File: documents/entityrelationship/model_test.go [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+		&config.Bootstrapper{},
+		&leveldb.Bootstrapper{},
+		&queue.Bootstrapper{},
+		&ideth.Bootstrapper{},
+		&configstore.Bootstrapper{},
+		anchors.Bootstrapper{},
+		documents.Bootstrapper{},
+		p2p.Bootstrapper{},
+		documents.PostBootstrapper{},
+		//&Bootstrapper{}, TODO activate bootstrapper for entity
```
Comment (mikiquantum) [inline documents/entityrelationship/model_test.go]: Do we need to enable this?

File: documents/entityrelationship/model_test.go [hunk 4]
```diff
@@ -0,0 +1,318 @@
...
+	m.GetTestCoreDocWithReset()
+
+	dr, err := m.CalculateDataRoot()
+	assert.NoError(t, err)
+	assert.False(t, utils.IsEmptyByteSlice(dr))
+}
+
+// TODO: proofs to support entity relationship prefixes
+func TestEntity_CreateProofs(t *testing.T) {
+	//e := createEntityRelationship(t)
```
Comment (mikiquantum) [inline documents/entityrelationship/model_test.go]: Can't we uncomment this test?

File: protobufs/entity/service.proto [hunk 5]
```diff
@@ -94,3 +94,22 @@ message EntityData {
   // Entity contact list
   repeated Contact contacts = 5;
 }
+
+message EntityRelationshipData {
+  string owner_identity = 1;
+  string label = 2;
```
Comment (mikiquantum) [inline protobufs/entity/service.proto]: I am wondering if we need this label at all here, as we will add more info with custom attributes in the near future.

File: documents/entityrelationship/model_test.go (hunk 3 above)
Comment (charleenfei) [reply documents/entityrelationship/model_test.go]: it's not been written yet

File: documents/entityrelationship/model.go [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+}
+
+// getClientData returns the entity relationship data from the entity relationship model
+func (e *EntityRelationship) getClientData() *cliententitypb.EntityRelationshipData {
+	var owner string
+	var target string
+	if e.OwnerIdentity != nil {
+		owner = e.OwnerIdentity.String()
+	}
+	if e.TargetIdentity != nil {
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]: use the converters in the identities for these conversions pls

File: documents/entityrelationship/model.go [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+}
+
+// createP2PProtobuf returns Centrifuge protobuf-specific EntityRelationshipData.
+func (e *EntityRelationship) createP2PProtobuf() *entitypb.EntityRelationship {
+	var didByte []byte
+	var tidByte []byte
+	if e.OwnerIdentity != nil {
+		didByte = e.OwnerIdentity[:]
+	}
+	if e.TargetIdentity != nil {
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]: same as above

File: documents/entityrelationship/model.go [hunk 8]
```diff
@@ -0,0 +1,319 @@
...
+
+	if data.TargetIdentity != "" {
+		if did, err := identity.NewDIDFromString(data.TargetIdentity); err == nil {
+			e.TargetIdentity = &did
+		} else {
+			return err
+		}
+	} else {
+		return identity.ErrMalformedAddress
+	}
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]: These blocks can be written with less if else using conversions in Identity

File: documents/entityrelationship/model.go [hunk 9]
```diff
@@ -0,0 +1,319 @@
...
+func (e *EntityRelationship) loadFromP2PProtobuf(entityRelationship *entitypb.EntityRelationship) error {
+	if entityRelationship.OwnerIdentity != nil {
+		did, err := identity.NewDIDFromBytes(entityRelationship.OwnerIdentity)
+		if err != nil {
+			return err
+		}
+		e.OwnerIdentity = &did
+	}
+
+	if entityRelationship.TargetIdentity != nil {
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]: same here

File: documents/entityrelationship/model.go [hunk 10]
```diff
@@ -0,0 +1,319 @@
...
+		tid, err := identity.NewDIDFromBytes(entityRelationship.TargetIdentity)
+		if err != nil {
+			return err
+		}
+		e.TargetIdentity = &tid
+	}
+
+	e.Label = entityRelationship.Label
+
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]: white space

File: documents/entityrelationship/model.go [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+	data, err := proto.Marshal(entityRelationship)
+	if err != nil {
+		return cd, errors.New("couldn't serialise EntityData: %v", err)
+	}
+
+	embedData := &any.Any{
+		TypeUrl: e.DocumentType(),
+		Value:   data,
+	}
+
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]: white space

File: documents/entityrelationship/model.go [hunk 12]
```diff
@@ -0,0 +1,319 @@
...
+	eProto := e.createP2PProtobuf()
+	if err != nil {
+		return nil, err
+	}
+	if e.CoreDocument == nil {
+		return nil, errors.New("getDocumentDataTree error CoreDocument not set")
+	}
+	t := e.CoreDocument.DefaultTreeWithPrefix(prefix, compactPrefix())
+	err = t.AddLeavesFromDocument(eProto)
+	if err != nil {
```
Comment (vedhavyas) [inline documents/entityrelationship/model.go]:
```suggestion
	if err := t.AddLeavesFromDocument(eProto); err != nil {
```

same for the next block as well

File: documents/entityrelationship/model_test.go [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+func (m *mockModel) ID() []byte {
+	args := m.Called()
+	id, _ := args.Get(0).([]byte)
+	return id
+}
+
+var testRepoGlobal documents.Repository
+
+func testRepo() documents.Repository {
+	if testRepoGlobal == nil {
```
Comment (vedhavyas) [inline documents/entityrelationship/model_test.go]: can be re-written to 
```
if testRepoGlobal != nil {
	return testRepoGlobal
}
```

**Comment by vedhavyas** [comment]
@charleenfei lets get the diff coverage from `62.89%.` to at least `80` or more

**Pull request merged**